frontend/js/actions/ArimaaActions.js

```diff
@@ -4,15 +4,101 @@ var APIUtils = require('../utils/WebAPIUtils.js');
+    if((data.meta.gUser !== undefined && data.meta.gUser.name == username) ||
+       (data.meta.sUser !== undefined && data.meta.sUser.name == username)) {
+      ArimaaActions.joinAndStartHeartbeatLoop(data.meta.gameID);
+    }
+    //Regardless of whether we're a player in this game or not, begin state update loop
+    ArimaaActions.startGameStateLoop();
+  },
+  initialStateError: function(data) {
+    ArimaaDispatcher.dispatch({
+      actionType: ArimaaConstants.ACTIONS.INITIAL_STATE_FAILED,
+      data: data
+    });
+  },
+
+  //Join game and heartbeat it in a loop
+  joinAndStartHeartbeatLoop: function(gameID) {
+    APIUtils.joinGame(gameID,
+                      function(data) {ArimaaActions.joinGameSuccess(gameID,data);},
+                      function(data) {ArimaaActions.joinGameError(gameID,data);});
+  },
+  joinGameSuccess: function(gameID, data) {
+    ArimaaDispatcher.dispatch({
+      actionType: ArimaaConstants.ACTIONS.GAME_JOINED,
+      gameID: gameID,
+      gameAuth: data.gameAuth
+    });
+    ArimaaActions.startHeartbeatLoop(gameID,data.gameAuth);
+  },
+  joinGameError: function(gameID, data) {
+    ArimaaDispatcher.dispatch({
+      actionType: ArimaaConstants.ACTIONS.GAME_JOIN_FAILED,
+      gameID: gameID,
+      data: data
+    });
+  },
+
+  //Initiates a loop heartbeating this game so long as this game is active and we're joined with it
+  startHeartbeatLoop: function(gameID, gameAuth) {
+    var username = UserStore.getUsername();
+    var game = ArimaaStore.getGameState();
 
+    //If the game is not active, terminate
+    if(game.meta.activeGameData === undefined)
+      return;
+
+    //If we're not joined to this game, terminate.
+    if((game.meta.gUser.name == username && !game.meta.activeGameData.gPresent) ||
+       (game.meta.sUser.name == username && !game.meta.activeGameData.sPresent))
+      return;
+
+    //TODO timeout
+    APIUtils.gameHeartbeat(gameID, gameAuth, FUNC_NOP, function (data) {ArimaaActions.onHeartbeatError(gameID,data);});
+     setTimeout(function () {ArimaaActions.startHeartbeatLoop(gameID, gameAuth);}, 5000);
+  },
+  onHeartbeatError: function(gameID,data) {
+    //TODO
+    console.log(data);
+    ArimaaDispatcher.dispatch({
+      actionType: ArimaaConstants.ACTIONS.HEARTBEAT_FAILED,
+      gameID: gameID
+    });
+  },
+
+  //Initiates a loop querying for the game state so long as the game is open or active.
+  startGameStateLoop: function() {
+    var game = ArimaaStore.getGameState();
+    //If the game is not open or active, terminate
+    if(game.meta.openGameData === undefined && game.meta.activeGameData === undefined)
+      return;
+    APIUtils.gameState(game.meta.gameID, game.meta.sequence+1, ArimaaActions.gameStateSuccess, ArimaaActions.gameStateError);
+  },
   gameStateSuccess: function(data) {
+    //TODO delete all the other stuff and just leave this action
+    //TODO why do we even need a dispatcher? Why not call methods directly on the ArimaaStore?
```

Comment:
The dispatcher keeps things simple since there could be multiple stores that we need to update. Right now they all go to ArimaaStore, but in theory we could add other stores.